### ISSUE TITLE
SqlDataAdapter.Fill very slow with Visual Studio 2022

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/SqlClient#1540]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/SqlDataAdapterFill-very-slow-with-Visua/1663322)._

---
[severity:I'm unable to use this version]
When running the project with Visual Studio 2019, we can run around 4500 SQL Queries per second, but with Visual Studio 2022 and the same code, we are down to 40(forty) IO Queries per second.
There is no performance difference between the debug and release mode.

An analysis with ReSharper shows that most of that time is used inside ntdll.dll.

---
#### Feedback Bot on 2/15/2022, 06:01 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by Wraith2** [comment]
Time used in ntdll.dll is probably just timers waiting on handles, the question is why you get more waiting now than you used to. Can you provide a minimal repro that I can run to see the problem myself?

**Comment by JRahnama** [comment]
@vsfeedback as @Wraith2 mentioned a small repro would be a good starting point.

**Comment by skurth** [comment]
I have the same issue when debugging. 

I have attached a sample project, but you have to use own database and adjust the sample project (search for "ToDo").

You can see the problem when you set a breakpoint at var sw = new Stopwatch() and on s = s. 
Compare the time from sw.ElapsedMilliseconds to the value shown in VS. In VS it says on my environment:

VS2019
Stopwatch (real world time): 690ms
VS debug time: 674ms 

VS2022
Stopwatch (real world time): **5023ms**
VS debug time: 859ms

If you need more diagnostic infos please let me know. Please also let me know if this is another issue and where I can post my issue.

[WindowsFormsApp1.zip](https://github.com/dotnet/SqlClient/files/8268078/WindowsFormsApp1.zip)

**Comment by Kaur-Parminder** [comment]
@skurth I am attaching console application that I have used for investigating this issue with Visual Studio 2019 and 2022 . The app make 5 subsequent attempts to do the query (only selecting 1 integer column to keep it simple for comparison). you can use Db.txt file to see  table (to avoid design/query difference).

I am highlighting the runs here for DB table in the attached App.

| VS2019 (Without Debug Mode)| VS2022 (Without Debug Mode)|
| 1257 | 962 |
| 453 | 348 |
| 354 | 340 |
| 323 | 347 |
| 369 | 359 |

|VS2019 (Debug Mode)| VS2022 (Debug Mode)|
| 780 | 806  |
| 355 |  527 |
| 400 | 442 |
| 491 |  338 |
| 462 | 382 |

 
[ConsoleApp1.zip](https://github.com/dotnet/SqlClient/files/8354389/ConsoleApp1.zip)
 
With multiple runs. I think on an average the SqlDataAdapter.Fill methods does fine with both VS versions. Also the run will show different results keeping in mind SQL Server Query Optimization.

**Comment by skurth** [comment]
@Kaur-Parminder With your example, I also can't reproduce the problem. It works fine in VS2019 and VS2022.

But when using my WinForms project and your database, the queries in VS2022 are up to 7x slower.
Please have a look at the following screenshots:

![VS2019_ms](https://user-images.githubusercontent.com/11660701/160437629-ce5d5f92-62d2-4d72-9dcd-09c248c37457.png)
![VS2022_ms](https://user-images.githubusercontent.com/11660701/160437632-edb146ec-2e27-4dd3-8ee2-3eebd3bd2cbe.png)

I have again attached my sample project where you can use your DB from ConsoleApp1 (you just need to configure the connection string). Press the button on Form1 to start the query execution.

[WindowsFormsApp2.zip](https://github.com/dotnet/SqlClient/files/8364406/WindowsFormsApp2.zip)

I'm not sure how this is related to WinForms. I have the same issue on a WCF service and Windows Service.

**Comment by Kaur-Parminder** [comment]
@skurth To avoid debug session, library/component loading setup, I have added message box to app that displays the elapsed time. I have attached copy of it. I am highlighting the runs from Windows Form App for both versions. 

| VS 2019 (Without Debug Mode) | VS 2022 (Without Debug Mode)
| 441| 416 |
| 285 | 322 |
| 263 | 309 |
| 242 | 256 |
| 260 | 312 |

| VS 2019 (With Debug Mode) | VS 2022 (With Debug Mode) |
| 7219 | 9650 |
| 8580 | 9520 |
| 8878 | 9250 |
| 8581 | 9106 |
| 9076 | 9759 |
[WindowsFormsApp2.zip](https://github.com/dotnet/SqlClient/files/8383971/WindowsFormsApp2.zip)

**Comment by skurth** [comment]
@Kaur-Parminder This are my results:

VS 2019 Debug
iteration 0 = 1155 ms
iteration 1 = 1060 ms
iteration 2 = 1092 ms
iteration 3 = 1016 ms
iteration 4 = 1007 ms

VS 2019 Without Debug
iteration 0 = 1073 ms
iteration 1 = 1078 ms
iteration 2 = 1050 ms
iteration 3 = 1069 ms
iteration 4 = 1013 ms

VS 2022 Debug
iteration 0 = 6678 ms
iteration 1 = 6102 ms
iteration 2 = 6188 ms
iteration 3 = 6178 ms
iteration 4 = 6138 ms

VS 2022 Without Debug
iteration 0 = 1100 ms
iteration 1 = 1010 ms
iteration 2 = 1098 ms
iteration 3 = 1008 ms
iteration 4 = 1183 ms

So it's only slow on VS 2022 and Debug Mode.
I've tested in on different machines with VS 2022 17.1.2

I tried to start VS 2022 in SafeMode, but when loading the project there is a "The object reference was not set to an object instance" and the project can't be loaded.

I also tried it with .NET 6 and Microsoft.Data.SqlClient. There is no problem. It's only an issue in .NET Framework. Results .NET 6 With Debug:
iteration 0 = 523 ms
iteration 1 = 599 ms
iteration 2 = 460 ms
iteration 3 = 462 ms
iteration 4 = 493 ms

After that, I've tried it with .NET Framework 4.8 again but changed from System.Data.SqlClient to Microsoft.Data.SqlClient. Results:
iteration 0 = 1078 ms
iteration 1 = 687 ms
iteration 2 = 458 ms
iteration 3 = 553 ms
iteration 4 = 487 ms

So I can only reproduce the problem when using:
- VS 2022
- Debug Mode
- System.Data.SqlClient

Any idea what could cause the issue?

**Comment by skurth** [comment]
I think I've found the problem/solution:
When comparing System.Data.SqlClient (slow) vs. Microsoft.Data.SqlClient in VS 2022, I can see the following difference:

![image](https://user-images.githubusercontent.com/11660701/161047909-8309555d-48d5-428e-ba0f-9cac69739fea.png)
![image](https://user-images.githubusercontent.com/11660701/161047972-8d042e5d-8049-478b-9fb0-4c75c3c0839b.png)

With System.Data.SqlClient (slow), there is an event shown in VS for every query execution:
It says (example):
Event Type Moniker	Event	Time	Duration	Thread
Execute Reader "SELECT PersonID FROM Persons WHERE PersonID = 721" The command text "SELECT PersonID FROM Persons WHERE PersonID = 721" was executed on connection "Data Source=XXX;Initial Catalog=YYY;Persist Security Info=True;User ID=ZZZ;Password=LLL;Encrypt=False", building a SqlDataReader. Hyperlink: Activate Historical Debugging	3.53s		[30892] Main Thread 

After that, I have disabled the following IntelliTrace options:
![image](https://user-images.githubusercontent.com/11660701/161049697-ca5a351d-a6be-4391-a160-3574b04b9bb0.png)

Now VS 2022 / System.Data.SqlClient runs as good as Microsoft.Data.SqlClient:
![image](https://user-images.githubusercontent.com/11660701/161049854-b361ccc9-850d-4f3b-8a74-27d6dfe1cb93.png)

(Note: In VS 2019 with those IntelliTrace options enabled and System.Data.SqlClient there are still no events shown -> Fast)

Summary:
VS 2019 / System.Data.SqlClient / IntelliTrace options enabled -> Fast (because no events are shown what could be wrong?)
VS 2019 / System.Data.SqlClient / IntelliTrace options disabled -> Fast
VS 2022 / System.Data.SqlClient / IntelliTrace options enabled -> Slow
VS 2022 / System.Data.SqlClient / IntelliTrace options disabled -> Fast
VS 2022 / Microsoft.Data.SqlClient / IntelliTrace options enabled -> Fast (because no events are shown what could be wrong?)
VS 2022 / Microsoft.Data.SqlClient / IntelliTrace options disabled-> Fast

I'm not sure if and where this needs to be fixed.
